<a href="https://colab.research.google.com/github/JeremyJerez/Introduction-to-Deep-Learning-and-Neural-Networks/blob/master/Week_5_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [2]:
#!pip install numpy
#!pip install pandas
#!pip install keras

In [3]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [5]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have.


In [6]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [7]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [8]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [9]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [10]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [11]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [13]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>


<a id='item32'></a>


## Import Keras


#### Let's go ahead and import the Keras library on a tensorflow backend


In [14]:
pip install tensorflow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [15]:
import keras

Let's import the rest of the packages from the Keras library that we will need to build our regression model.


In [16]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [17]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has one hidden layers with 10 neurons.


Let's import scikit-learn in order to randomly split the data into a training and test sets

In [18]:
from sklearn.model_selection import train_test_split 


Splitting the data into a training and test sets by holding 30% of the data for testing


In [19]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)


## Train and Test the Network


### Creating the model


In [20]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [21]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 116874.6484
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 47686.2969
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 14261.8584
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 3780.7551
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 2847.6284
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 2533.7126
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 2266.8206
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 2037.2479
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1835.2059
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1659.7970
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1506.2319
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 1369.9185
Epoch 13/

Model evaluation


In [22]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val


10/10 [==============================] - 0s 2ms/step


185.1510009765625

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.



Let's import the mean_squared_error function from Scikit-learn.



In [23]:
from sklearn.metrics import mean_squared_error



In [24]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)


185.1510127901012 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.


In [26]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 91.1156005859375
10/10 [==============================] - 0s 2ms/step
MSE 2: 106.47694396972656
10/10 [==============================] - 0s 2ms/step
MSE 3: 107.30651092529297
10/10 [==============================] - 0s 2ms/step
MSE 4: 111.06465911865234
10/10 [==============================] - 0s 2ms/step
MSE 5: 118.62484741210938
10/10 [==============================] - 0s 2ms/step
MSE 6: 111.32294464111328
10/10 [==============================] - 0s 2ms/step
MSE 7: 129.69911193847656
10/10 [==============================] - 0s 2ms/step
MSE 8: 106.61772918701172
10/10 [==============================] - 0s 2ms/step
MSE 9: 124.51341247558594
10/10 [==============================] - 0s 2ms/step
MSE 10: 107.36573028564453
10/10 [==============================] - 0s 2ms/step
MSE 11: 100.59979248046875
10/10 [==============================] - 0s 2ms/step
MSE 12: 92.58605194091797
10/10 [==============================] - 0s 2ms/step
MSE 13: 103.68976593017578
10/10 [==================